kuksa-client/kuksa_client/grpc/__init__.py

```diff
@@ -576,10 +573,14 @@ def _prepare_get_request(self, entries: Iterable[EntryRequest]) -> val_pb2.GetRe
         logger.debug("%s: %s", type(req).__name__, req)
         return req
 
-    def _process_get_response(self, response: val_pb2.GetResponse) -> List[DataEntry]:
+    def _process_get_response(self, response: val_pb2.GetResponse) -> (List[DataEntry], List[str]):
```

Comment:
I don't think this change is a good idea.

It breaks every client using the `kuksa.grpc` library.
It creates a very weird interface. Previously, when doing:
```python
client.get_current_values(["Vehicle.Speed", "Vehicle.Services"])
```

you would either get:
* a list of entries
* an exception (for example permission denied)

with this change, you would either get:
* a list of entries
* a list of entries and a list of errors
* an exception (for example permissions denied)

The previous behaviour was a lot more pythonic, and easier to work with.

I guess the main culprit, is that the the grpc function should either be successful and return the expected result **or** fail and return an error. Otherwise, I think the end result will be exceptionally confusing.

Reply:
I would agree that "half-errored" is maybe unexpected. If we don't want that, I think the main question is, when there is a problem with "some" entries, do we want to
 - just return an error
 - silenty ignore the problematic cases and continue with the working ones

Next question, are these two questions
 - Using a wildcard
 - Listing all entries a wildcard would match by hand
 equivalent in that regard, or should we treat one differently from the other?

Opinions?

Reply:
The only way I can think of for now, assuming we don't want to break anything by modifying ```kuksa.grpc``` would be John's suggestion today during the daily.

We disable erros for wildcards. That leads to the following:

- If the request is wildcard the response contains only allowed entries (silenty ignore the problematic cases)
- If the request is for a signal path without permission an error is returned.
- If the request is a list of entries (with and without permission), only the error permission is returned. The client should be aware that it cannot query that entry and should modify the request.

Reply:
Yes, maybe.

I'm having trouble foreseeing all the implications. The "just return an error" would probably be the best in terms of having a predictable behaviour.

But something a long the line of what @rafaeling listed above might be more pragmatic in some sense, but would require implementing special handling which would probably make it harder to understand and predict (as a user).

I can easily imagine how "silently ignoring errors" for wildcards might lead to unintended behaviour in clients.

The fundamental issue, I believe, is that what can be requested is very unrestricted. You can request:
* paths together with wildcard paths.
* different fields for every path or wildcard path.

And in combination with a global error + a global (optional) list of per path errors (next to the successful results), it is not at all clear how this _should_ behave.

Reply:
I also see how silently ignoring errors can greatly confuse clients. Another option that could work and could be to return just an error for each wildcard (would be treated as signal path), while also including a permission field for each signal in the metadata. This way, the clients are fully aware upfront of what they have access to or not.

The client should start by calling ```getMetaData``` to retrieve the entire tree with permissions included and then build the requests from there. Additionally, this feature could allow our kuksa-client to, for instance, add colors to differentiate between entries with permissions and those without when using autocompletion.

Reply:
> I also see how silently ignoring errors can greatly confuse clients. Another option that could work could be to return just an error for each wildcard (would be treated as signal path), while also including a permission field for each signal in the metadata.

So, if I understand correctly, that would be "return only error", but there would be one error per "path" requested.

> while also including a permission field for each signal in the metadata. This way, the clients are fully aware upfront of what they have access to or not.

That's an interesting idea.

Reply:
To follow up on my comment above:
> The fundamental issue, I believe, is that what can be requested is very unrestricted. You can request:
> * paths together with wildcard paths.
> * different fields for every path or wildcard path.

and dig into why this is problematic.

### Mixing "listing" of signals with "getting" signals.
As identified above, there is the question of whether wildcards should be handled differently than when requesting explicit paths. If the answer is yes, the ability to combine paths and wildcards in the same request is causing issues. One way to solve this is to have separate functions for these separate behaviours, e.g.
* `get(...)`  (get explicitly listed signals)
* `list(...)` (list available signals)

The new function `list(...)` would then list all signals that is available to be listed (it would be the "wildcard handling" function).
The `get(...)` function would return an error if permission is denied for an explicitly listed signal. Or (since it would be a 1:1 relationship between the number of paths requested and the number returned, it could also return a list with either a successful response or an error, i.e. `[(path, success_response), (path, success_response), (path, error), (path, success_response)]`)


### Mixing retrieval of metadata with retrieval of data
The other point is that both metadata and data can be requested for the same path. That leads to the question of what should happen if there is permission to read one of them but not the other.

One way to solve this would be to have different functions for these things.
* `get_values(...)` (and `list_values(...)`)
* `get_metadata(...)` (and/or `list_metadata(...)`)

`get_values(...)` would return an error if there is no permission to read the value for an explicitly listed path. Since it would be a 1:1 relationship between the number of paths requested and the number returned, it could also return a list with either a successful response or an error, i.e. `[(path, success_response), (path, success_response), (path, error), (path, success_response)]`)
`get_metadata(...)` or `list_metadata(...)` would not fail based on lacking permissions for the actual data.

Reply:
I think the metadata idea is nice, but let's decouple it form the discussion here.

So I created this token

```json
{
  "sub": "local dev",
  "iss": "createToken.py",
  "aud": [
    "kuksa.val"
  ],
  "iat": 1516239022,
  "exp": 1767225599,
  "scope": "read:Vehicle.Acceleration.Lateral read:Vehicle.Acceleration.Vertical"
}
```


and running this example:

```python
from kuksa_client.grpc import EntryRequest
from kuksa_client.grpc import Field
from kuksa_client.grpc import View
from kuksa_client.grpc import VSSClient

with VSSClient('127.0.0.1', 55556) as client:
    path = 'Vehicle.Body.Windshield.Front.Wiping.System.TargetPosition'
    client.authorize("eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJsb2NhbCBkZXYiLCJpc3MiOiJjcmVhdGVUb2tlbi5weSIsImF1ZCI6WyJrdWtzYS52YWwiXSwiaWF0IjoxNTE2MjM5MDIyLCJleHAiOjE3NjcyMjU1OTksInNjb3BlIjoicmVhZDpWZWhpY2xlLkFjY2VsZXJhdGlvbi5MYXRlcmFsIHJlYWQ6VmVoaWNsZS5BY2NlbGVyYXRpb24uVmVydGljYWwifQ.00PTCsHWocFEn5PjFqfvUwVo400WleuR4IK7SJu8MDCi0MZe19Pg3BFjCKefybQfwxgHpR_vuwxFtQPigvWTSVbQFMdn1zbOBulF9x_g7KGBjVcLNkw3MOQU5eswM-9e6rRyj5qWuL97tczOK21f3QEVsGacHWO9wYvPmw-Fb4PWOJUrx82ans7iuMPt3SCFjPa_of-NAPotgA14ahXHRzfLrlLFtHmgIeDthNGt8Fd-wPd5sl8r2vb7XQq1-eOb-v01nTwp2WFXwa3ZJV_EegNZzvM8Wf4nJl8zfRLNfmIqHlvpcaOnIVIHUBsfteJS4dICZhs6Ub3xscnEEEF9g37hIrQkLpXa96K4YcaxDmJr_MQH3qfhYX2Q0U_74KPxKzDv7fL5LYsLUB2RQ2114rxyD4jRVPheNwB56ytjLBDXTGuCZvVLpxouhCIn7cFfoSf24djy65BOhD7vFAE_ET1fIqvycNEl83o6QkKjKuLvJwC6omHlD1cTG99nOJBlNj3fJer4hw8iFy3xQjSE2eh0WVqeEmD7FjpfnlauovvSOndK5Qrk_Ih89OypvrgOqPE2n9EQuQ-UgHVB7ip2fU1Mo1GzpdxXgn7luRIx4K4fKESzhWrI90nkuIh2hb7pKhYWy8jIxrh2m2nSWld8Dr9Y9MCoYxd0uWoA7gxLASQ")
    entries = client.get_current_values([
        'Vehicle.Acceleration.Vertical', 'Vehicle.Acceleration.Longitudinal', 'Vehicle.Acceleration.Lateral'
    ])  

    print(str(entries))
```

This yields

```
 python test.py
Traceback (most recent call last):
  File "/Users/scs2rng/Documents/Dev/testkc/test.py", line 9, in <module>
    entries = client.get_current_values([
  File "/Users/scs2rng/Documents/Dev/kuksa.val/kuksa-client/kuksa_client/grpc/__init__.py", line 727, in get_current_values
    entries = self.get(
  File "/Users/scs2rng/Documents/Dev/kuksa.val/kuksa-client/kuksa_client/grpc/__init__.py", line 679, in wrapper
    return func(self, *args, **kwargs)
  File "/Users/scs2rng/Documents/Dev/kuksa.val/kuksa-client/kuksa_client/grpc/__init__.py", line 889, in get
    return self._process_get_response(resp)
  File "/Users/scs2rng/Documents/Dev/kuksa.val/kuksa-client/kuksa_client/grpc/__init__.py", line 581, in _process_get_response
    self._raise_if_invalid(response)
  File "/Users/scs2rng/Documents/Dev/kuksa.val/kuksa-client/kuksa_client/grpc/__init__.py", line 644, in _raise_if_invalid
    raise VSSClientError(error, errors)
kuksa_client.grpc.VSSClientError: ({'code': 403, 'reason': 'forbidden', 'message': 'Permission denied'}, [{'path': 'Vehicle.Acceleration.Longitudinal', 'error': {'code': 403, 'reason': 'forbidden', 'message': 'Permission denied'}}])
```

That was my baseline, and I think this is definitely one valid choice: My intention was "get values for three paths", the API telling me "Nope, that is not possible, here is one hint as to why"

When I modify this exmaple to 

```python
from kuksa_client.grpc import EntryRequest
from kuksa_client.grpc import Field
from kuksa_client.grpc import View
from kuksa_client.grpc import VSSClient

with VSSClient('127.0.0.1', 55556) as client:
    path = 'Vehicle.Body.Windshield.Front.Wiping.System.TargetPosition'
    client.authorize("eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJsb2NhbCBkZXYiLCJpc3MiOiJjcmVhdGVUb2tlbi5weSIsImF1ZCI6WyJrdWtzYS52YWwiXSwiaWF0IjoxNTE2MjM5MDIyLCJleHAiOjE3NjcyMjU1OTksInNjb3BlIjoicmVhZDpWZWhpY2xlLkFjY2VsZXJhdGlvbi5MYXRlcmFsIHJlYWQ6VmVoaWNsZS5BY2NlbGVyYXRpb24uVmVydGljYWwifQ.00PTCsHWocFEn5PjFqfvUwVo400WleuR4IK7SJu8MDCi0MZe19Pg3BFjCKefybQfwxgHpR_vuwxFtQPigvWTSVbQFMdn1zbOBulF9x_g7KGBjVcLNkw3MOQU5eswM-9e6rRyj5qWuL97tczOK21f3QEVsGacHWO9wYvPmw-Fb4PWOJUrx82ans7iuMPt3SCFjPa_of-NAPotgA14ahXHRzfLrlLFtHmgIeDthNGt8Fd-wPd5sl8r2vb7XQq1-eOb-v01nTwp2WFXwa3ZJV_EegNZzvM8Wf4nJl8zfRLNfmIqHlvpcaOnIVIHUBsfteJS4dICZhs6Ub3xscnEEEF9g37hIrQkLpXa96K4YcaxDmJr_MQH3qfhYX2Q0U_74KPxKzDv7fL5LYsLUB2RQ2114rxyD4jRVPheNwB56ytjLBDXTGuCZvVLpxouhCIn7cFfoSf24djy65BOhD7vFAE_ET1fIqvycNEl83o6QkKjKuLvJwC6omHlD1cTG99nOJBlNj3fJer4hw8iFy3xQjSE2eh0WVqeEmD7FjpfnlauovvSOndK5Qrk_Ih89OypvrgOqPE2n9EQuQ-UgHVB7ip2fU1Mo1GzpdxXgn7luRIx4K4fKESzhWrI90nkuIh2hb7pKhYWy8jIxrh2m2nSWld8Dr9Y9MCoYxd0uWoA7gxLASQ")

    entries = client.get_current_values([
        'Vehicle.Acceleration.*'
    ])  

    print(str(entries))
```
I get with current master

```
{'Vehicle.Acceleration.Longitudinal': None, 'Vehicle.Acceleration.Vertical': None, 'Vehicle.Acceleration.Lateral': None}
```

that is not good, have not tested this branch, but my _expectation_ would be, that this behaves exactly as if I had requested all matching paths "by hand".

That would seem consistent to me.

The suggested metadata extension is than one interesting way to get around this for "generic"/"exploratory" apps, but as I said that is an interesting suggestion anyway, apart from the current issue

Reply:
> that is not good, have not tested this branch, but my expectation would be, that this behaves exactly as if I had requested all matching paths "by hand".

That particular response is caused by the "permissions bug" in the first implementation of wildcard support.